catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/DumpCommand.java

```diff
@@ -407,13 +412,134 @@ protected void printStatus(long count) {
     return metacardTransformerList;
   }
 
-  private Optional<QueryResponseTransformer> getZipCompression() throws InvalidSyntaxException {
-    if (!zipCompression.isPresent()) {
-      zipCompression =
-          getServiceByFilter(
-              QueryResponseTransformer.class,
-              "(|" + "(" + Constants.SERVICE_ID + "=" + ZIP_COMPRESSION + ")" + ")");
+  private void createZip(SourceResponse upstreamResponse, String filePath)
+      throws CatalogTransformerException {
+    if (upstreamResponse.getResults() == null || upstreamResponse.getResults().isEmpty()) {
+      throw new CatalogTransformerException("Source response cannot be null or empty");
+    }
+
+    try (FileOutputStream fileOutputStream = new FileOutputStream(filePath);
+        ZipOutputStream zipOutputStream = new ZipOutputStream(fileOutputStream)) {
+
+      List<Result> resultList = upstreamResponse.getResults();
+      Map<String, Resource> resourceMap = new HashMap<>();
+
+      // write the metacards to the zip
+      resultList
+          .stream()
+          .map(Result::getMetacard)
+          .forEach(
+              metacard -> {
+                writeMetacardToZip(zipOutputStream, metacard);
+
+                if (hasLocalResources(metacard)) {
+                  resourceMap.putAll(getAllMetacardContent(metacard));
+                }
+              });
+
+      // write the resources to the zip
+      resourceMap.forEach(
+          (filename, resource) -> writeResourceToZip(zipOutputStream, filename, resource));
+
+    } catch (IOException e) {
+      throw new CatalogTransformerException(
+          String.format(
+              "Error occurred when initializing/closing ZipOutputStream with path %s.", filePath),
+          e);
+    }
+  }
+
+  private void writeMetacardToZip(ZipOutputStream zipOutputStream, Metacard metacard) {
+
+    try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        ObjectOutputStream objectOutputStream = new ObjectOutputStream(byteArrayOutputStream)) {
+
+      ZipEntry zipEntry = new ZipEntry(METACARD_PATH + metacard.getId());
+      zipOutputStream.putNextEntry(zipEntry);
+
+      objectOutputStream.writeObject(new MetacardImpl(metacard));
+      InputStream inputStream = new ByteArrayInputStream(byteArrayOutputStream.toByteArray());
```

Comment:
❗️ ✏️ I think this call to `byteArrayOutputStream.toByteArray()` can still blow up memory. I would recommend using guava's [FileBackedOutputStream](https://google.github.io/guava/releases/18.0/api/docs/com/google/common/io/FileBackedOutputStream.html), and calling `.toByteSource().openStream()`.
```
  private void writeMetacardToZip(ZipOutputStream zipOutputStream, Metacard metacard) {
    InputStream inputStream = null;

    try (FileBackedOutputStream fileBackedOutputStream = new FileBackedOutputStream(THRESHOLD);
        ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileBackedOutputStream)) {

      ZipEntry zipEntry = new ZipEntry(METACARD_PATH + metacard.getId());
      zipOutputStream.putNextEntry(zipEntry);

      objectOutputStream.writeObject(new MetacardImpl(metacard));
      inputStream = fileBackedOutputStream.asByteSource().openStream();

      IOUtils.copy(inputStream, zipOutputStream);
    } catch (IOException e) {
      LOGGER.debug("Failed to add metacard with id {}.", metacard.getId(), e);
    } finally {
      try {
        inputStream.close();
      } catch (IOException | NullPointerException ignore) {
      }
    }
  }
```